Finite stuff

angeris opened the pull request:
A bit of a mammoth one, unfortunately, since there's a lot of moving parts to being able to define "cardinality of a finite set" appropriately.

Tl;dr, the rough game plan for the current PR:
- From before, `FiniteSet` is a set which is contained in some `List`
- One simple way of defining cardinality of a `s: FiniteSet` is to use the existence of this containing `List` (call it `list`) and then 'minimize' the list size: `list.filter(s.contains).unique`
- Consequences: need to prove that this is indeed the smallest possible list which contains a given set (i.e., any other list which contains a given set is at least as large in `.length`)
- This then gets into a bunch of other hairy things and supporting cast members (including, to make life slightly easier, the addition of a `remove_elem` function for a `List`) which are necessary to prove that, e.g., two lists with the same `.contains` have the same `.length` (by removing items one-by-one until there are none in both `List`s)

There is a question of whether to continue going down the rabbit hole, or simply wait for dependent types (for which the easiest notion of cardinality is something like 'there exists an injection' `FinRange n -> FiniteSet<T>` or whatever and we can later prove that this very easily matches the definition of `list.unique.length` in `List`).

Draft for now, since there's a bunch of open questions and some of the theorems are incomplete (since either I haven't finished them, such as `eq_contains_implies_unique_eq_length` or I can't get the prover to be happy/recognize induction, as in `remove_element_removes_no_other_elements`)

Also the proofs are unfortunately not super pretty imo, since they require a bunch of proofs by induction.

lacker commented:
Interesting. I think this approach seems okay, induction proofs just really are long, just kind of boilerplatey. It might help to break out just the theorems about list and get those validated in a separate PR, in the long run there will probably be many other parts of math that indirectly rely on list theorems because humans tend to use those informally a lot.

One point I'm definitely taking away from this is that the current `.yaml` setup for storing the cache files is super annoying. I'm hoping to improve that as part of the current cache refactoring...

angeris commented:
Ok, fixed proofs... now we can finally define cardinality of a `FiniteSet`!

Agree re: separating the commits to `List` and `FiniteGroup`. Will do that (likely will make a separate PR here).

**A general sidenote:** One thing I need to get used to, which is a good workflow probably to add to docs, is to _write the goal at the end of the scope at all times_. Sometimes the prover needs a lot of help, but sometimes I don't even notice that it can already prove the final claim at the end of the scope without my help, but intermediate steps I've added to "help" it in the final goal require additional work.

angeris commented:
Yeah, ok, it's devolved a little bit. There's finally enough `List` theorems here (and equivalences, and definitions, etc) to prove something interesting, namely that every element in a `FiniteGroup` (defined in another branch, which I will PR once this one is accepted) has finite order!

lacker commented:
This is great! We really don't need smaller commits for too much - it does become unwieldy to deal with files when they become too long, but that's partly due to the speed of the verifier, and I'd rather improve verifier speed than complain about file size in advance.
